modules/rostests/apitests/user32/ScrollBarRedraw.c

```diff
@@ -0,0 +1,712 @@
+
+static BOOL HaveHRedraw = FALSE;
+static BOOL HaveVRedraw = FALSE;
+
+typedef enum
+{
+    FSM_STATE_START,
+    FSM_STATE_VSCR_SHOWN,
+    FSM_STATE_VSCR_HIDDEN,
+    FSM_STATE_HSCR_SHOWN,
+    FSM_STATE_HSCR_HIDDEN,
+    FSM_STATE_BSCR_SHOWN,
+    FSM_STATE_BSCR_HIDDEN,
+    FSM_STATE_WIDTH_SHRUNK,
+    FSM_STATE_WIDTH_EXPANDED,
+    FSM_STATE_HEIGHT_SHRUNK,
+    FSM_STATE_HEIGHT_EXPANDED,
+    FSM_STATE_BOTH_SHRUNK,
+    FSM_STATE_BOTH_EXPANDED,
+    FSM_STATE_END
+} FSM_STATE;
+
+#define FSM_STEP_PERIOD_MS 250
+
+static UINT_PTR FsmTimer = 0;
+static UINT CurrentColor = 0;
+static FSM_STATE FsmState = FSM_STATE_START;
+
+static int ClientWidth = 0;
+static int ClientHeight = 0;
+
+static int OrigWidth = 0;
+static int OrigHeight = 0;
+static int SmallWidth = 0;
+static int SmallHeight = 0;
+
+static void ColorsCleanup()
+{
+    UINT Iter;
+
+    for (Iter = 0; Iter < TEST_COLOR_COUNT; Iter++)
+    {
+        if (ColorBrushes[Iter] != NULL)
+        {
+            DeleteObject(ColorBrushes[Iter]);
+            ColorBrushes[Iter] = NULL;
+        }
+    }
+}
+
+static BOOL ColorsInit(void)
+{
+    UINT Iter;
+
+    if (TEST_COLOR_COUNT != 16)
+    {
+        return FALSE;
+    }
+
+    /* Standard Windows 16-Color VGA Color palette. */
+    Colors[0]  = RGB(0x00, 0x00, 0x00);  /* Black */
+    Colors[1]  = RGB(0x00, 0x00, 0x80);  /* Dark Blue */
+    Colors[2]  = RGB(0x00, 0x80, 0x00);  /* Dark Green */
+    Colors[3]  = RGB(0x00, 0x80, 0x80);  /* Dark Cyan */
+    Colors[4]  = RGB(0x80, 0x00, 0x00);  /* Dark Red */
+    Colors[5]  = RGB(0x80, 0x00, 0x80);  /* Dark Magenta */
+    Colors[6]  = RGB(0x80, 0x80, 0x00);  /* Dark Yellow */
+    Colors[7]  = RGB(0xC0, 0xC0, 0xC0);  /* Light Gray */
+    Colors[8]  = RGB(0x80, 0x80, 0x80);  /* Dark Gray */
+    Colors[9]  = RGB(0x00, 0x00, 0xFF);  /* Blue */
+    Colors[10] = RGB(0x00, 0xFF, 0x00);  /* Green */
+    Colors[11] = RGB(0x00, 0xFF, 0xFF);  /* Cyan */
+    Colors[12] = RGB(0xFF, 0x00, 0x00);  /* Red */
+    Colors[13] = RGB(0xFF, 0x00, 0xFF);  /* Magenta */
+    Colors[14] = RGB(0xFF, 0xFF, 0x00);  /* Yellow */
+    Colors[15] = RGB(0xFF, 0xFF, 0xFF);  /* White */
+
+    for (Iter = 0; Iter < TEST_COLOR_COUNT; Iter++)
+    {
+        ColorBrushes[Iter] = CreateSolidBrush(Colors[Iter]);
+        if (ColorBrushes[Iter] == NULL)
+        {
+            ColorsCleanup();
+            return FALSE;
+        }
+    }
+
+    return TRUE;
+}
+
+static void RunTestWindow(const char *ClassName, const char *WindowTitle, UINT ClassStyle)
+{
+    WNDCLASS Class = { 0 };
+    HWND Window;
+    MSG  Message;
+    ATOM ClassAtom;
+
+    CurrentColor = 0;
+
+    Class.style         = ClassStyle;
+    Class.lpfnWndProc   = WindowProc;
+    Class.cbClsExtra    = 0;
+    Class.cbWndExtra    = 0;
+    Class.hInstance     = HINST_THISCOMPONENT;
+    Class.hIcon         = LoadIcon(NULL, IDI_APPLICATION);
+    Class.hCursor       = LoadCursor(NULL, IDC_ARROW);
```

Comment:
```suggestion
    Class.hIcon         = LoadIconW(NULL, IDI_APPLICATION);
    Class.hCursor       = LoadCursorW(NULL, IDC_ARROW);
```

Reply:
Should I switch everything to unicode?  The only piece of text that I have in this testcase is the window caption and class name, and those are currently ANSI strings.

Reply:
Actually, I don't think selecting the API type is correct here, since the type of the IDI_/IDC_ constants will depend on whether UNICODE is defined or not, whereas the versions with an API suffix require the inputs to be of a specific type.

And since the return value is a handle, I think that the natural thing to do is to use these calls without a suffix, to match the type of the IDI_/IDC_ constants.

Though I guess there is potentially value in using the API suffix, to trigger -Werror in case the the state of the UNICODE macro doesn't match the code.

Reply:
> Should I switch everything to unicode? The only piece of text that I have in this testcase is the window caption and class name, and those are currently ANSI strings.

I was going to suggest that, actually (we prefer having the explicit unicode stuff)

Reply:
> > Should I switch everything to unicode? The only piece of text that I have in this testcase is the window caption and class name, and those are currently ANSI strings.
> 
> I was going to suggest that, actually (we prefer having the explicit unicode stuff)

Okay, I've replaced all of the API calls I could with explicit Unicode variants...  With the exception of these two lines, I've kept them on the default, since they need to be the ANSI variant when compiled under user32_apitest.exe, otherwise it would fail on Werror, but if the user were to switch the compilation mode to Unicode, an explicit ANSI variant would cause Werror on the Unicode build.

If you'd like me to use the explicit ANSI versions, please let me know.

Reply:
Thanks for confirming that I can use the compilation mode versions of those two functions via Mattermost!